File: uva.onlinejudge.org/uva 455 - Periodic Strings/455.py

"""这个python提交结果runtime error, cpp文件ok"""

def NextIndex(last_index, line):
    c = line[0]
    for i in range(last_index+1, len(line)):
        if line[i] == c:
            return i

    return len(line)

def Check(index, line):
    length = index
    repeat = 1
    while (repeat * length) < len(line):
        for i in range(length):
            if length + i >= len(line):
                return False
            
            if line[i] != line[repeat*length + i]:
                return False
        repeat += 1
        
    return True

def Solve(line):
    last_index = 0
    found = False
    while (not found) and (last_index < len(line)-1):
        last_index = NextIndex(last_index, line)
        found = Check(last_index, line)
        if found:
            print(last_index)
            return
    
    print(len(line))
    
def main():
    lines = int(input())
    
    for i in range(lines):
        if i:
            print()
            
        line = input()
        while not line:
            line = input()
        
        Solve(line)
        

if __name__ == '__main__':
    main()

File: uva.onlinejudge.org/uva 455 - Periodic Strings/455.cpp
/*
 * 455.cpp
 *
 *  Created on: Apr 8, 2017
 *      Author: frank
 */

#include <iostream>
#include <string>
#include <stdint.h>

using namespace std;

size_t FindNext(size_t last_index, const string& line)
{
	for(size_t i = last_index + 1; i <= line.size()/2; ++i)
	{
		//cout<<"|||||||||||||||||"<<i<<"\n";
		if(line[0] == line[i] && line.size() % i == 0)
			return i;
	}

	return line.size();
}

bool Check(size_t index, const string& line)
{
	size_t i = 1;
	while(i * index < line.size())
	{
		for(size_t n = 0; n < index; ++n)
		{
			if(line[n] != line[i*index+n])
				return false;
		}

		++i;
	}

	return true;
}

void Solve(const string& line)
{
	size_t last_index = 0;
	bool found = false;
	while(!found && last_index <= line.size()/2)
	{
		//cout<<":::::"<<last_index<<"\n";
		last_index = FindNext(last_index, line);
		found = Check(last_index, line);
		if(found)
		{
			cout<<last_index<<"\n";
			return;
		}
	}

	cout<<line.size()<<"\n";
}

int main()
{
	int32_t count;
	cin>>count;
	string line;
	for(int32_t i = 0; i < count; ++i)
	{
		if(i > 0)
		{
			cout<<"\n";
		}

		cin>>line;
		//cout<<"=====>"<<line<<"\n";
		Solve(line);
	}
}

